src/compose/service-manager.ts

```diff
@@ -32,626 +32,618 @@ type ServiceManagerEventEmitter = StrictEventEmitter<
-				setTimeout(() => this.listenToEvents(), 1000);
-			});
+		return new Promise((resolve, reject) => {
+			parser
+				.on('error', (e: Error) => {
+					log.error('Error on docker events stream:', e);
+					reject(e);
+				})
+				.on('end', resolve);
+			stream.pipe(parser);
+		});
+	};
+
+	Bluebird.resolve(listen())
+		.catch((e) => {
+			log.error('Error listening to events:', e, e.stack);
+		})
+		.finally(() => {
+			listening = false;
+			setTimeout(() => listenToEvents(), 1000);
+		});
 
-		return;
-	}
+	return;
+}
 
-	public async attachToRunning() {
-		const services = await this.getAll();
-		for (const service of services) {
-			if (service.status === 'Running') {
-				const serviceId = service.serviceId;
-				const imageId = service.imageId;
-				if (serviceId == null || imageId == null) {
-					throw new InternalInconsistencyError(
-						`serviceId and imageId not defined for service: ${service.serviceName} in ServiceManager.start`,
-					);
-				}
+export async function attachToRunning() {
+	const services = await getAll();
+	for (const service of services) {
+		if (service.status === 'Running') {
+			const serviceId = service.serviceId;
+			const imageId = service.imageId;
+			if (serviceId == null || imageId == null) {
+				throw new InternalInconsistencyError(
+					`serviceId and imageId not defined for service: ${service.serviceName} in ServiceManager.start`,
+				);
+			}
 
-				if (service.containerId == null) {
-					throw new InternalInconsistencyError(
-						`containerId not defined for service: ${service.serviceName} in ServiceManager.attachToRunning`,
-					);
-				}
-				logger.attach(service.containerId, {
-					serviceId,
-					imageId,
-				});
+			if (service.containerId == null) {
+				throw new InternalInconsistencyError(
+					`containerId not defined for service: ${service.serviceName} in ServiceManager.attachToRunning`,
+				);
 			}
+			logger.attach(service.containerId, {
+				serviceId,
+				imageId,
+			});
 		}
 	}
+}
 
-	public async getContainerIdMap(appId: number): Promise<Dictionary<string>> {
-		return _(await this.getAllByAppId(appId))
-			.keyBy('serviceName')
-			.mapValues('containerId')
-			.value() as Dictionary<string>;
-	}
+export async function getContainerIdMap(
+	appId: number,
+): Promise<Dictionary<string>> {
+	return _(await getAllByAppId(appId))
+		.keyBy('serviceName')
+		.mapValues('containerId')
+		.value() as Dictionary<string>;
+}
 
-	private reportChange(containerId?: string, status?: Partial<Service>) {
-		if (containerId != null) {
-			if (status != null) {
-				this.volatileState[containerId] = {};
-				_.merge(this.volatileState[containerId], status);
-			} else if (this.volatileState[containerId] != null) {
-				delete this.volatileState[containerId];
-			}
+function reportChange(containerId?: string, status?: Partial<Service>) {
+	if (containerId != null) {
+		if (status != null) {
+			volatileState[containerId] = {};
+			_.merge(volatileState[containerId], status);
```

Comment:
I think this can be
```suggestion
			volatileState[containerId] = { ...status };
```
since there's nothing to merge into. And if you don't need the cloning side-effect then just `= status` would work instead